File: linux-monitor-backend/app/models/monitor.py
from pydantic import BaseModel


class SystemInfo(BaseModel):
    hostname: str
    platform: str
    uptime: int
    loadAverage: list[float]
    timestamp: int


class CpuInfo(BaseModel):
    usage: float
    cores: list[float]
    frequency: float
    temperature: float | None = None
    timestamp: int


class MemoryInfo(BaseModel):
    total: int
    used: int
    available: int
    percent: float
    swap: dict
    timestamp: int


class DiskInfo(BaseModel):
    total: int
    used: int
    free: int
    percent: float
    readSpeed: float
    writeSpeed: float
    timestamp: int


class NetworkInterface(BaseModel):
    name: str
    isUp: bool
    speed: int
    mtu: int
    ipAddress: str
    macAddress: str
    bytesReceived: int
    bytesSent: int
    packetsReceived: int
    packetsSent: int
    errorsReceived: int
    errorsSent: int


class NetworkConnections(BaseModel):
    tcp: int
    udp: int
    tcpListen: int
    tcpEstablished: int
    tcpTimeWait: int


class OpenPort(BaseModel):
    port: int
    protocol: str
    status: str
    process: str | None = None
    pid: int | None = None


class NetworkInfo(BaseModel):
    uploadSpeed: float
    downloadSpeed: float
    totalSent: int
    totalReceived: int
    connections: NetworkConnections
    interfaces: list[NetworkInterface]
    openPorts: list[OpenPort]
    timestamp: int


class ProcessInfo(BaseModel):
    pid: int
    name: str
    cpuPercent: float
    memoryPercent: float
    status: str


class MonitorData(BaseModel):
    system: SystemInfo
    cpu: CpuInfo
    memory: MemoryInfo
    disk: DiskInfo
    network: NetworkInfo
    processes: list[ProcessInfo]

File: linux-monitor-dashboard/src/types/monitor.ts
export interface SystemInfo {
  hostname: string;
  platform: string;
  uptime: number;
  loadAverage: number[];
  timestamp: number;
}

export interface CpuInfo {
  usage: number;
  cores: number[];
  frequency: number;
  temperature?: number;
  timestamp: number;
}

export interface MemoryInfo {
  total: number;
  used: number;
  available: number;
  percent: number;
  swap: {
    total: number;
    used: number;
    percent: number;
  };
  timestamp: number;
}

export interface DiskInfo {
  total: number;
  used: number;
  free: number;
  percent: number;
  readSpeed: number;
  writeSpeed: number;
  timestamp: number;
}

export interface NetworkInterface {
  name: string;
  isUp: boolean;
  speed: number; // Mbps
  mtu: number;
  ipAddress: string;
  macAddress: string;
  bytesReceived: number;
  bytesSent: number;
  packetsReceived: number;
  packetsSent: number;
  errorsReceived: number;
  errorsSent: number;
}

export interface NetworkConnections {
  tcp: number;
  udp: number;
  tcpListen: number;
  tcpEstablished: number;
  tcpTimeWait: number;
}

export interface OpenPort {
  port: number;
  protocol: 'tcp' | 'udp';
  status: 'listening' | 'established' | 'time_wait';
  process?: string;
  pid?: number;
}

export interface NetworkInfo {
  uploadSpeed: number;
  downloadSpeed: number;
  totalSent: number;
  totalReceived: number;
  connections: NetworkConnections;
  interfaces: NetworkInterface[];
  openPorts: OpenPort[];
  timestamp: number;
}

export interface ProcessInfo {
  pid: number;
  name: string;
  cpuPercent: number;
  memoryPercent: number;
  status: string;
}

export interface MonitorData {
  system: SystemInfo;
  cpu: CpuInfo;
  memory: MemoryInfo;
  disk: DiskInfo;
  network: NetworkInfo;
  processes: ProcessInfo[];
}